nemo/collections/asr/parts/submodules/fast_rnnt_greedy_decoding.py

```diff
@@ -0,0 +1,431 @@
+
+                    self.not_all_blank_t.copy_(~self.blank_mask.all())
+                    self.symbols_added_t += 1
+                    self.seq_idx_t += 1
+
+                self.symbols_per_time_step.index_copy_(0, self.time_idx_t, self.symbols_added_t)
+                self.time_idx_t += 1
+
+            self.scores_cpu.copy_(self.scores.transpose(0, 1).contiguous(), non_blocking=True)
+            self.labels_cpu.copy_(self.labels.transpose(0, 1).contiguous(), non_blocking=True)
+            self.symbols_per_time_step_cpu.copy_(self.symbols_per_time_step, non_blocking=True)
+
+            self.last_label.fill_(self.caller._SOS)
+            self.time_idx_t.fill_(0)
+
+            (self.graph,) = cu_call(cudart.cudaStreamEndCapture(torch.cuda.current_stream().cuda_stream))
+        (self.graph_exec,) = cu_call(cudart.cudaGraphInstantiate(self.graph, 0))
+
+    def __call__(
+        self,
+        caller,
+        x: torch.Tensor,
+        out_len: torch.Tensor,
+        device: torch.device,
+        partial_hypotheses: Optional[List[rnnt_utils.Hypothesis]] = None,
+    ):
+
+        # Need to copy x and out_len into "staging buffers", or do a graph update.
+
+        if partial_hypotheses is not None:
+            raise NotImplementedError("`partial_hypotheses` support is not supported")
+
+        assert not caller.preserve_alignments
+        assert not caller.preserve_frame_confidence
+
+        batch_size = x.shape[0]
+        # ideally we would use out_len.max() here...
+        max_time = x.shape[1]
+
+        if torch.is_autocast_enabled():
+            x = x.to(torch.get_autocast_gpu_dtype())
+
+        # What do we do if batch_size is actually smaller for the
+        # input? Is this a problem? the clone() call will fail...
+        if max_time > self.max_time or batch_size > self.batch_size:
+            self._reinitialize(max_time, batch_size, x, out_len)
+
+        torch.cuda.nvtx.range_push("Graph exec")
+        self.encoder_output[: x.shape[0], : x.shape[1], ...].copy_(x)
+        self.encoder_output_length[: out_len.shape[0]].copy_(out_len)
+        cu_call(cudart.cudaGraphLaunch(self.graph_exec, torch.cuda.current_stream().cuda_stream))
+        cu_call(cudart.cudaStreamSynchronize(torch.cuda.current_stream().cuda_stream))
+        torch.cuda.nvtx.range_pop()
+
+        torch.cuda.nvtx.range_push("Copy data out")
+
+        torch.cuda.nvtx.range_push("GPU->CPU out_len copy")
+        out_len_cpu = out_len.cpu()
+        torch.cuda.nvtx.range_pop()
+
+        hypotheses = [
+            rnnt_utils.Hypothesis(score=0.0, y_sequence=[], timestep=[], dec_state=None) for _ in range(batch_size)
+        ]
+
+        # At batch size 16, this for loop takes 25-31 milliseconds... How can we speed it up?
+        # It is likely that most of the overhead is in memory loads...
+        # Let's profile just this block
+        for i in range(batch_size):
+            j = 0
+            for t in range(out_len_cpu[i]):
```

Comment:
This can be potentially done on GPU
If I correctly understand what's going on here, the algorithm can be the following:
- for each sequence find first non-blank label index, possibly with torch.arange + mask, see https://stackoverflow.com/a/60202801
- make a mask for resulting symbols
- reshape (flatten) and take non-masked elements
- convert to list (`Tensor.tolist`)